### ISSUE TITLE
dart2js: Deferred library loading improvements

### ISSUE THREAD

**Issue by rakudrama** [dart-lang/sdk#29686]
Observations on a timeline of an app.

The app loads ~50 parts that total 9.5MB to load a deferred library.
The timeline is 1370s from the `loadLibrary()` call to being able to execute code.
620ms is spend requesting the ~50 part.js files
Each part file puts a top-level JS anonymous function in a table (indexed by a hash to allow the table to be shared by many separately compiled applications.)
There are four outstanding requests at any one time. Chrome allows 6 per domain, and there are two data requests outstanding during this process, leaving four for JS downloads. (There are more outstanding requests with https.)
After 620ms, all the onLoaded events have fired and completed the Futures for the requests.
At this point all the functions are called in a tight loop.
749ms is spent calling the functions, of which 349ms is spent inside the functions doing initialization.
The missing 400ms is accounted for by the first full-parse of the functions.
In the 620ms of the download waterfall, the CPU is underutilized.

Observations

- Requests are delayed by the browser so that there are only four outstanding. Reducing the number of requests would reduce this resource saturation latency.

- The requests have a per-request latency and a data-length dependent latency. Small files are much more expensive per byte than larger files to download, but the parsing and initialization are proportional to length.

- Appending small files to the initial download would reduce requests and help the above two points

- Merging parts, or putting several parts in one file, would reduce the protocol/server latency contributing to the fixed part of a request.

- Parts could be structured so that there are more stages to the evaluation, and some could be done earlier. The work could be structured so that the initial function needs to be called to acquire the function that is called when all the parts are download. This initial call would force some parsing. This could be overlapped with downloading. If all the parsing could be forced early, the 400ms could be overlapped with the 620ms downloading, reducing the total 1370ms to 970ms.

- Parts could be downloaded in an order that increases overlapping. e.g. it might be better to load the biggest part first rather than last.

- It might be beneficial to split large parts into smaller parts to allow more kinds of work to be moved into phases that can happen early. If the cost of running parts of the initialization can be reduced below a janky threshold, e.g. 4ms, then it can be evaluated pro-actively during 'interactive' use rather than only after a loadLibrary() request.

Suggestions

- [ ] Place several small hunks in one _part_ file. Make download logic distinguish between hunks and files.
- [ ] Add a stage for 'parsing' that does order-independent and context-independent actions. Make loadDeferredLibrary call the parsing entry when the download is available.
- [ ] Extend the order-independent actions - e.g. call some methods twice, the first time exits early but forces parsing.
- [ ] Sort hunks by containing file size so larger files are requested earlier (TODO: experiment with other schedules)

**Comment by sigmundch** [comment]
Other ideas:
 [ ] Create tools to post-process the emitted hunks and cluster in custom ways. This could be specified with a yaml input file as a specification of how code should be clustered.

**Comment by sigmundch** [comment]
* [ ] reduce the boilerplate: small/empty chunks have too much unnecessary boilerplate.

**Comment by Aetet** [comment]
@rakudrama @sigmundch  Hi. Any news about that?

**Comment by sigmundch** [comment]
Not much has happened in this area yet, since most of our focus right now is in getting dart2js to support Dart 2.0 semantics.

One change that did happen was that @rakudrama made it more compact to represent the table of what output units are composed of. This mostly reduces the codesize of the main output unit a little bit: https://github.com/dart-lang/sdk/commit/31cd4e145744a46d6ff773edfff470898c9ea87a

**Comment by rakudrama** [comment]
https://github.com/dart-lang/sdk/commit/2a3a5c88c7eddb41801c89473c1bdae649391d60 removes empty part files, but only under `--fast-startup`.

**Comment by robbecker-wf** [comment]
I'd also like dart2js to be more helpful in identifying bad practices and how to improve. Specifically, identifying that a part file is too big, and provide some guidance or clues to where in the code to add a deferred boundary to reduce the part size. 
These recommendations could also take into account data from the Chrome JS coverage output to determine how much of a part file is actually used when loaded and suggest splitting out the parts that aren't used.
